src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -13,472 +13,528 @@ namespace nbl::asset
+                    if (descInfo.stride == 0u)
+                        return false;
+                }
+                else
                 {
-                    auto& last = r->back();
-                    if (members.array[i].offset == (last.offset + last.size))
-                        last.size += members.array[i].size;
-                    else
-                        r->push_back({shaderStage, members.array[i].offset, members.array[i].size});
+                    descInfo.count = std::accumulate(stageIntroBindingInfo.count.begin(), stageIntroBindingInfo.count.end(), 1u, multiplyArrayDimensions);
+                    descInfo.stride = descInfo.count;
+                    if (descInfo.stride == 0u)
+                        return false;
                 }
             }
-            introspection++;
-            r++;
+            else
+            {
+                descInfo.count = 1u;
+                descInfo.stride = 0u;
+            }
+
+            const auto& pplnIntroDataFoundBinding = m_descriptorSetBindings[i].find(descInfo);
+            if (pplnIntroDataFoundBinding != m_descriptorSetBindings[i].end())
+            {
+                if (pplnIntroDataFoundBinding->type != stageIntroBindingInfo.type)
+                    return false;
+                if (pplnIntroDataFoundBinding->count != descInfo.count)
+                    return false;
+                if (pplnIntroDataFoundBinding->stride != descInfo.stride)
+                    return false;
+            }
+
+            debug << "binding ID: " << descInfo.binding << "type: " << static_cast<uint32_t>(descInfo.type) << std::endl;
+
+            descriptorsToMerge[i].insert(descInfo);
         }
+        std::cout << debug.str() << std::endl;
     }
 
-    core::vector<SPushConstantRange> merged;
-
-    SPushConstantRange rngToPush; rngToPush.offset = 0u; rngToPush.size = 0u;
-    uint32_t stageFlags = 0u;
-    for (uint32_t i = 0u; i < ICPUMeshBuffer::MAX_PUSH_CONSTANT_BYTESIZE/sizeof(uint32_t); i += sizeof(uint32_t))
+    // validate if only descriptors with the highest bindings are run-time sized
+    std::array<int32_t, ICPUPipelineLayout::DESCRIPTOR_SET_COUNT> highestBindingsTmp = m_highestBindingNumbers;
+    for (uint32_t i = 0u; i < ICPUPipelineLayout::DESCRIPTOR_SET_COUNT; ++i)
     {
-        SPushConstantRange curr; curr.offset = i; curr.size = sizeof(uint32_t);
-
-        uint32_t tmpFlags = 0u;
-        for (uint32_t stg=0u; stg<infos.size(); stg++)
-            for (const SPushConstantRange& rng : ranges[stg])
-                if (curr.overlap(rng))
-                    tmpFlags |= rng.stageFlags;
+        const auto& introBindingInfos = stageData->getDescriptorSetInfo(i);
 
-        if (!i)
-            stageFlags = tmpFlags;
+        for (const auto& stageIntroBindingInfo : introBindingInfos)
+            highestBindingsTmp[i] = std::max<const int32_t>(highestBindingsTmp[i], stageIntroBindingInfo.binding);
```

Comment:
you could continually increase this value while doing `descriptorsToMerge[i].insert(descInfo);` in the loop above, saves you two iterations